### ISSUE TITLE
C++: memory leaks on some exceptions

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#214]
Current code generated by CppCompiler is not really exception-safe. It's structured in a way that implies that `_read()` called from constructor always completes in full, and that's not always true if an exception occurs in the middle of the process.

For example, in `if_struct.ksy`:

```cpp
if_struct_t::if_struct_t(kaitai::kstream *p_io, kaitai::kstruct* p_parent, if_struct_t *p_root) : kaitai::kstruct(p_io) {
    m__parent = p_parent;
    m__root = this;
    _read();
}

void if_struct_t::_read() {
    m_op1 = new operation_t(m__io, this, m__root);
    m_op2 = new operation_t(m__io, this, m__root);
    m_op3 = new operation_t(m__io, this, m__root);
}

if_struct_t::~if_struct_t() {
    delete m_op1;
    delete m_op2;
    delete m_op3;
}
```

If one of `new operation_t` would throw an exception (for example, because of not enough bytes to read and/or other reasons, such as IO error, or basically any other possible C++ exception), this would result in constructor failing → destructor called to everything clean up → some of `m_opX` members would be contain garbage → `delete` would be called on them → it would most likely result in a segfault.

The correct way to structure this code would be probably:

1. Clean up all "dangerous" (i.e. pointer) members before any exceptions might occur
2. Wrap all pointer deletions into safeguard if: `if (m_op1)  delete m_op1;`

**Comment by scottmsilver** [comment]
My view is that  it's worth resolving the debate on smart pointers before doing this. If you were to chose the smart pointer path the clean-up paths are a priori exception safe.

**Comment by webbnh** [comment]
As I mentioned in #414, I've just traced a significant memory leak in my project to this issue.

For reference, here's a snippet of my KSY:
```
  payload:
    seq:
    - id: data
      process: xor(_parent.mask_field & 0xff)
      size-eos: true
      type:
        cases:
# ...
          rec_types::record_3: record_3
# ...
        switch-on: _parent.type
```

And here's the corresponding snippet from the generated code:
```
void myrecords_t::payload_t::_read() {
    n_data = true;
    switch (_parent()->type()) {
// ...
    case REC_TYPES_RECORD_3: {
        n_data = false;
        m__raw__raw_data = m__io->read_bytes_full();
        m__raw_data = kaitai::kstream::process_xor_one(m__raw__raw_data, (_parent()->mask_field() & 255));
        m__io__raw_data = new kaitai::kstream(m__raw_data);
        m_data = new record_3_t(m__io__raw_data, this, m__root);
        break;
    }
// ...
    default: {
        m__raw__raw_data = m__io->read_bytes_full();
        m__raw_data = kaitai::kstream::process_xor_one(m__raw__raw_data, (_parent()->mask_field() & 255));
        break;
    }
    }
}
```

(By the way, it looks like you could hoist the contents of the `default` case and place them before the the `switch`.)

What is happening in my case is that the substream for Record 3 is too small, and so the attempt to read it inside its constructor throws an exception which, because we've already allocated the `m__io__raw_data`, leaks that block.

While I would certainly appreciate KS going over to using smart pointers (I'm using C++11), I understand and support the aspiration to support a wide variety of versions (even if I think it's a little crazy... ;-) ).  But, I'm wondering if there aren't some other options available.

For instance, if the c'tor were to be generated with member initialization for all the fields like so (which I'm led to understand is "best practice", anyway):
```
myrecords_t::payload_t::payload_t(kaitai::kstream* p__io, myrecords_t::record_t* p__parent, myrecords_t* p__root) :
    kaitai::kstruct(p__io),
    m_data(nullptr),
    n_data(true),
    m__root(p__root),
    m__parent(p__parent),
    m__raw__raw_data(),
    m__raw_data(),
    m__io__raw_data(nullptr)
{
    _read();
}
```

Then, in `_read()` we are free to delete everything we expect to allocate -- either it will actually have been allocated (in which case, in the event of an exception, it should be deleted), or it will be a null pointer, which is an [OK thing to try to delete](https://en.cppreference.com/w/cpp/memory/new/operator_delete):
> if ptr is a null pointer, the standard library deallocation functions do nothing

So, then my read function code would become:
```
    case REC_TYPES_RECORD_3: {
        m__raw__raw_data = m__io->read_bytes_full();
        m__raw_data = kaitai::kstream::process_xor_one(m__raw__raw_data, (_parent()->mask_field() & 255));
        try {
            m__io__raw_data = new kaitai::kstream(m__raw_data);
            m_data = new record_3_t(m__io__raw_data, this, m__root);
        } catch (const std::exception& e) {
            delete m__io__raw_data;
            delete m_data;
            throw e;
        }
        n_data = false;
        break;
    }
```
It fetches and processes the data (and exceptions thrown by that transfer control out before we try to allocate anything).  It then tries to do the two allocations:  if either one fails, we delete them both and then re-throw the original exception; otherwise, control leaves the `try` block and clears `n_data` (although, if you embrace the deletion of null pointers concept, you should be able to get rid of `n_data` altogether).

Does that seem reasonable?

**Comment by generalmimon** [comment]
@GreyCat 
> If one of `new operation_t` would throw an exception (for example, because of not enough bytes to read and/or other reasons, such as IO error, or basically any other possible C++ exception), this would result in **constructor failing → destructor called** to everything clean up

I think that this isn't the case. I don't know C++ much to confirm it, but various sources on the internet claim that the destructor is not called if the constructor throws an exception. The reason is that if the constructor refuses to create an instance, the object is considered to be uninitialized, so there's nothing to destruct. From https://wiki.c2.com/?MemoryLeakInCpp:

> Suppose you want to exit a constructor without creating an instance. The only way I know of is to throw an exception, (...). This, however, has a memory leak because **destructors are never called** if the **constructor** throws an exception.

**Comment by generalmimon** [comment]
@GreyCat 
> The correct way to structure this code would be probably:
> 
> 1. Clean up all "dangerous" (i.e. pointer) members before any exceptions might occur
> 2. Wrap all pointer deletions into safeguard if: `if (m_op1)  delete m_op1;`

I agree. As I see it, the structure of the C++98 code in auto-read mode could look like this:

```cpp
if_struct_t::if_struct_t(kaitai::kstream* p__io, kaitai::kstruct* p__parent, if_struct_t* p__root) : kaitai::kstruct(p__io) {
    m__parent = p__parent;
    m__root = this;
    m_op1 = 0;
    m_op2 = 0;
    m_op3 = 0;
    try {
        _read();
    } catch(...) {
        this->~if_struct_t();
        throw;
    }
}

void if_struct_t::_read() {
    m_op1 = new operation_t(m__io, this, m__root);
    m_op2 = new operation_t(m__io, this, m__root);
    throw "error";
    m_op3 = new operation_t(m__io, this, m__root);
}

if_struct_t::~if_struct_t() {
    if (m_op1) delete m_op1;
    if (m_op2) delete m_op2;
    if (m_op3) delete m_op3;
}
```

Note that in the auto-read mode, it's necessary to call the destructor manually from the `catch` block of the constructor, because we already know that the constructor will throw an exception, so the destructor will be never called automatically.

When auto-read is disabled, the `_read()` won't be called from the constructor, so the constructor can't fail with an exception -> we don't need to call the destructor manually (and not even the application code that calls `_read()`, it should be called automatically), but as we know that it eventually gets called at some point, the `if` safeguards must remain so that we clean only the stuff that got initialized.

In C++11 standard, all owned things by an KS type use smart pointers now and the destructors are empty, so it doesn't have to care about any manual destruction.

**Comment by generalmimon** [comment]
@webbnh 
> Then, in `_read()` we are free to delete everything we expect to allocate -- either it will actually have been allocated (in which case, in the event of an exception, it should be deleted), or it will be a null pointer, which is an [OK thing to try to delete](https://en.cppreference.com/w/cpp/memory/new/operator_delete):
> 
> > if ptr is a null pointer, the standard library deallocation functions do nothing

What you write here is interesting. If it works even in the C++98 mode (i.e. if `int *pi = 0;` is a null pointer), then it could mean that we don't even need the `if (m_op1)` safeguards at all (because we are already pre-initializing all owning pointers to null - see https://github.com/kaitai-io/kaitai_struct/issues/244). I'll try to omit the safeguards and see if it works.

**Comment by generalmimon** [comment]
Here's another article to this topic: http://jsteemann.github.io/blog/2015/11/18/on-exception-handling/

**Comment by generalmimon** [comment]
@generalmimon:

> I agree. As I see it, the structure of the C++98 code in auto-read mode could look like this:

> If it works even in the C++98 mode (i.e. if `int *pi = 0;` is considered a null pointer), then it could mean that we don't even need the `if (m_op1)` safeguards at all (because we are already pre-initializing all owning pointers to null - see #244). I'll try omitting the safeguards and see if it works.

Well, unfortunately I made it even worse. This is the diff between the [`cpp_stl_98/clang3.5_linux`](https://github.com/kaitai-io/ci_artifacts/blob/cpp_stl_98/clang3.5_linux/test_out/cpp_stl_98/ci.json) target output `ci.json` [before](https://github.com/kaitai-io/ci_artifacts/blob/2fa57910db520e08e0270f9f8964848ef3af2c70/test_out/cpp_stl_98/ci.json) and [after](https://github.com/kaitai-io/ci_artifacts/blob/535ca81ddae19a8b26ccf4630a9176044e6719c8/test_out/cpp_stl_98/ci.json) the change from https://github.com/kaitai-io/kaitai_struct_compiler/commit/44f090f788f75edf47e81714b27d54944577af22 (I just filtered the `elapsed` key from both `ci.json` versions so we can see some relevant stuff):


<details>
<summary>ci.json diff</summary>

```diff
diff --git 1/ci_cpp-try-catch-a.json 2/ci_cpp-try-catch-b.json
index 0f28049..596f1b4 100644
--- 1/ci_cpp-try-catch-a.json
+++ 2/ci_cpp-try-catch-b.json
@@ -4,7 +4,13 @@
         "is_kst": true
     },
     "ValidFailEqBytes": {
-        "status": "passed",
+        "status": "failed",
+        "failure": {
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
+            "trace": null
+        },
         "is_kst": true
     },
     "ParamsPassArrayUsertype": {
@@ -43,11 +49,11 @@
         "is_kst": true
     },
     "DefaultEndianExprException": {
-        "status": "leak",
+        "status": "failed",
         "failure": {
-            "file_name": null,
-            "line_num": null,
-            "message": "286 (24 direct, 262 indirect) bytes in 1 blocks are definitely lost in loss record 7 of 9",
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
             "trace": null
         },
         "is_kst": true
@@ -88,7 +94,13 @@
         "is_kst": true
     },
     "ValidFailRangeStr": {
-        "status": "passed",
+        "status": "failed",
+        "failure": {
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
+            "trace": null
+        },
         "is_kst": true
     },
     "RepeatUntilComplex": {
@@ -189,7 +201,13 @@
         "is_kst": true
     },
     "EofExceptionU4": {
-        "status": "passed",
+        "status": "failed",
+        "failure": {
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
+            "trace": null
+        },
         "is_kst": true
     },
     "Expr3": {
@@ -348,7 +366,13 @@
         "is_kst": true
     },
     "ValidFailContents": {
-        "status": "passed",
+        "status": "failed",
+        "failure": {
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
+            "trace": null
+        },
         "is_kst": true
     },
     "StrLiterals": {
@@ -608,11 +632,11 @@
         "is_kst": true
     },
     "EosExceptionBytes": {
-        "status": "leak",
+        "status": "failed",
         "failure": {
-            "file_name": null,
-            "line_num": null,
-            "message": "415 (384 direct, 31 indirect) bytes in 1 blocks are definitely lost in loss record 8 of 9",
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
             "trace": null
         },
         "is_kst": true
@@ -643,11 +667,11 @@
         "is_kst": true
     },
     "EosExceptionU4": {
-        "status": "leak",
+        "status": "failed",
         "failure": {
-            "file_name": null,
-            "line_num": null,
-            "message": "415 (384 direct, 31 indirect) bytes in 1 blocks are definitely lost in loss record 9 of 9",
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
             "trace": null
         },
         "is_kst": true
@@ -744,7 +768,13 @@
         "is_kst": true
     },
     "ValidFailRangeBytes": {
-        "status": "passed",
+        "status": "failed",
+        "failure": {
+            "file_name": "unknown location",
+            "line_num": "0",
+            "message": "signal: SIGABRT (application abort requested)",
+            "trace": null
+        },
         "is_kst": true
     },
     "SwitchManualStr": {
@@ -868,13 +898,13 @@
     },
     "$meta": {
         "lang": "cpp_stl",
-        "timestamp": "2020-07-30T22:55:01Z",
+        "timestamp": "2020-07-31T20:32:08Z",
         "ci": {
             "ci": "travis",
-            "build_id": "713497026",
-            "job_id": "713497028",
-            "job_number": "384.2",
-            "url": "https://travis-ci.org/kaitai-io/ci_targets/jobs/713497028"
+            "build_id": "713821539",
+            "job_id": "713821541",
+            "job_number": "385.2",
+            "url": "https://travis-ci.org/kaitai-io/ci_targets/jobs/713821541"
         }
     }
 }
```
</details>

This is what can be found in [`results.xml`](https://github.com/kaitai-io/ci_artifacts/blob/535ca81ddae19a8b26ccf4630a9176044e6719c8/test_out/cpp_stl_98/results.xml) about `default_endian_expr_exception` for instance:

```xml
<TestCase name="test_default_endian_expr_exception">
    <Exception file="unknown location" line="0"><![CDATA[signal: SIGABRT (application abort requested)]]>
        <LastCheckpoint file="/home/travis/build/kaitai-io/ci_targets/tests/spec/cpp_stl_98/test_default_endian_expr_exception.cpp" line="17"><![CDATA[]]></LastCheckpoint>
    </Exception>
    <TestingTime>206</TestingTime>
</TestCase>
```

That mentions the line [`tests/spec/cpp_stl_98/test_default_endian_expr_exception.cpp:17`](https://github.com/kaitai-io/kaitai_struct_tests/blob/9fc2a0eef1b9e88e60f562cb5ff7d2962419af6f/spec/cpp_stl_98/test_default_endian_expr_exception.cpp#L17).

Can the `SIGABRT` signal mean a segmentation fault? It would make a lot of sense to me if the destructor is called twice: once manually in the [`catch` block](https://github.com/kaitai-io/ci_targets/blob/a9298ebf72c8d8954b62c22057eb20358ecc2a5b/compiled/cpp_stl_98/default_endian_expr_exception.cpp#L15) inside the constructor, and for the second time automatically by the C++ compiler (or the Boost library? hm, doesn't really matter) on the line reported in the `report.xml` as [_LastCheckpoint_](https://github.com/kaitai-io/kaitai_struct_tests/blob/9fc2a0eef1b9e88e60f562cb5ff7d2962419af6f/spec/cpp_stl_98/test_default_endian_expr_exception.cpp#L17).

That would mean that we really need the safeguards, so I'll try adding them.

**Comment by generalmimon** [comment]
Well, safeguards are [added](https://github.com/kaitai-io/ci_targets/commit/5ef260714097f21f31726a6e77c13c1c2683caa4), but they changed absolutely nothing, so now I don't know what to do. What else can `SIGABRT` mean, if not segfault? Does anybody know?

Cc @GreyCat, @webbnh, @ams-tschoening, @jchv

**Comment by webbnh** [comment]
`SIGABRT` is the result of calling the `abort()` C RTL function.  There might be lots of things that do that.  The likely ones are unhandled exceptions and blown assertions (the latter is probably reported via an exception mechanism), so, yes, it could be a `SEGV` which is causing the `SIGABRT` (particularly if something tries to catch it and then either fumbles or rethrows it).

I'm not sorry that adding the safeguards didn't change things, since I'm hoping that they aren't necessary or useful.  :-/

I would think that if you ran the failing test case under a debugger it would be able to point you to the source of the problem.  (I don't know how easy it is to do that....)

**Comment by ams-tschoening** [comment]
Initializing with `nullptr` only WITHOUT safeguards is NOT enough. Look at the following code in use before:

	default_endian_expr_exception_t::default_endian_expr_exception_t(kaitai::kstream* p__io, kaitai::kstruct* p__parent, default_endian_expr_exception_t* p__root) : kaitai::kstruct(p__io) {
	    m__parent = p__parent;
	    m__root = this;
	    m_docs = 0;

	    try {
	        _read();
	    } catch(...) {
	        this->~default_endian_expr_exception_t();
	        throw;
	    }
	}

	default_endian_expr_exception_t::~default_endian_expr_exception_t() {
	    for (std::vector<doc_t*>::iterator it = m_docs->begin(); it != m_docs->end(); ++it) {
	        delete *it;
	    }
	    delete m_docs;
	}

While the first thing `_read` does is allocating a vector for `m_docs`, if that fails for any reason, e.g. because one is low on memory or else, the DTOR will access a `nullptr` for `m_docs` and fail. So in case of a `nullptr`, this NEEDS a safeguard.

OTOH, if safeguards are in place already, setting things to `nullptr` after deleting them make sense as well. This prevents them from being freed by accident one additional time. It simply wouldn't matter if the DTOR would be called twice, while it currently does and the safeguards wouldn't change anything regarding that. Calling `delete` doesn't set things to a `nullptr`, so the safeguards will always succeed, which in case of `m_docs` in the DTOR above will result in accessing invalid memory IF the DTOR is called a second time for any reason.

Could a DTOR be called more than once with the current implementation? Possibly [YES](https://stackoverflow.com/a/51119036/2055163)! Things heavily depend on how parent objects maintain members and how members of the throwing instance are used. Just think of the example above again: If some instance of `doc_t` throws, the vector `m_docs` would still be iterated to free other instances of that type. While in this example the throwing instance shouldn't be part of that vector, because `push_back` wouldn't succeed, there simply might be other cases where the same instance IS freed multiple times.

So, to summarize things: When accessing a vector like `m_docs` in the DTOR above, safeguards ARE NECESSARY. If instance variables have been deleted, `NULL` them afterwards as well to make safeguards work properly in case of unexpected multiple recalls.

**Comment by Mingun** [comment]
Why just do not use C++11 generator and implement missing stuff of C++11 standard in C++98 runtime?

**Comment by generalmimon** [comment]
@ams-tschoening Thanks a lot for your explanation and suggestion!

> **Calling `delete` doesn't set things to a `nullptr`**, so the safeguards will always succeed, which in case of `m_docs` in the DTOR above will result in accessing invalid memory IF the DTOR is called a second time for any reason.

Thanks for this point, I didn't know that, it seemed logical to me that `delete` could work that way 🙂

As I see it, the main point of the safeguards was exactly to make the destructor immune to multiple calls, so it totally makes sense to fix them.

It makes a lot of sense now, if the destructor is for some reason called twice, that could in theory cause the errors `Invalid read of size 4: Address 0x616d0a0 is 16 bytes inside a block of size 27 free'd` and immediately afterwards `Invalid free() / delete / delete[] / realloc(): Address 0x616d090 is 0 bytes inside a block of size 27 free'd` that [Valgrind reports](https://github.com/kaitai-io/ci_artifacts/blob/1865176c89f243f386e9a036bcd2f01d356721f5/test_out/cpp_stl_98/valgrind.xml#L337-L660) for the source [`default_endian_expr_exception.cpp`](https://github.com/kaitai-io/ci_targets/blob/5ef260714097f21f31726a6e77c13c1c2683caa4/compiled/cpp_stl_98/default_endian_expr_exception.cpp).

So I'll try `NULL`ing the deleted pointers and we'll see if it helps.

@Mingun:

> Why just do not use C++11 generator and implement missing stuff of C++11 standard in C++98 runtime?

Please open a new issue with this very ambitious and long-term plan (or maybe comment to https://github.com/kaitai-io/kaitai_struct/issues/211, though I'm not sure if that's the right place).

Right now (i.e. before 0.9 release), I would like to just solve all known C++ memory leaks preferably in the easiest and fastest way (but of course, the code must remain safe and robust) in both C++98 and C++11 mode to ensure that Kaitai Struct C++ codegen is usable even for projects like Kismet (https://github.com/kaitai-io/kaitai_struct/issues/747). @GreyCat said on Gitter that it should be done just before that we release 0.9 (see https://gitter.im/kaitai_struct/Lobby?at=5ecec7a4b101510b201f7193). And I'd like to release 0.9 later this month, if it's possible, not in a year or two if we get consumed by big issues.

**Comment by generalmimon** [comment]
> So I'll try `NULL`ing the deleted pointers and we'll see if it helps.

Unfortunately, again absolutely no change. File `ci.json` looks [still the same](https://github.com/kaitai-io/ci_artifacts/commit/9cba84b5142d225efa8433acf2ee29fada2f5036#diff-e7c28116f07c958735759a1ba3e9a86d), as well as [`valgrind.xml`](https://github.com/kaitai-io/ci_artifacts/commit/9cba84b5142d225efa8433acf2ee29fada2f5036#diff-24a6622688daab02c0eee7a2f8daff07).

What a shame that I don't have Linux and fully working C++ debugger in it, that could really help to pinpoint the problem. Any chance that 

I have gone through the [Valgrind stack traces](https://github.com/kaitai-io/ci_artifacts/blob/9cba84b5142d225efa8433acf2ee29fada2f5036/test_out/cpp_stl_98/valgrind.xml#L337-L660) several times, but they don't make much sense to me. The line numbers of the last-known location often lead to a line with only the closing curly bracket `}` of a CTOR or DTOR, which I don't know what to conclude from it.

Maybe it means that the `clang` compiler tries to clean some memory automatically on that closing curly bracket? If so, which one and how we can avoid the error?

**Comment by ams-tschoening** [comment]
> Maybe it means that the clang compiler tries to clean some memory automatically on that closing curly bracket? If so, which one and how we can avoid the error?

While I don't understand the details as well yet, I suggest focusing on the following for some time:

> <fn>std::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;::~basic_string()</fn>

That call should belong to the field `std::string m_indicator;` and because that is NOT a pointer, it's automatically cleaned up by the runtime. That might exactly be why e.g. line 62 is logged, which is `}` in the DTOR. Cleaning the string might exactly happen there as the last statement.

In case of an exception the DTOR is called manually and might clean this string up. BUT in case of an exception the CTOR is left with an exception as well and CLANG might simply clean the SAME string up a second time! That would happen at line 51 of the CTOR of `doc_t` and [exactly that line is logged](https://github.com/kaitai-io/ci_artifacts/blob/9cba84b5142d225efa8433acf2ee29fada2f5036/test_out/cpp_stl_98/valgrind.xml#L358) as well.

This looks to me like two times attempting to free the SAME string because CLANG isn't prepared that the DTOR has been called manually already. If I understand this thread correctly, introducing the manual DTOR-call is exactly what made things worse and triggers this problem.

So comment the DTOR-call ONLY in compiled code and leave everything else like it is. If I'm correct, things should succeed again with only logging memory leaks.

**Comment by generalmimon** [comment]
@ams-tschoening:

> So comment the DTOR-call ONLY in compiled code and leave everything else like it is. If I'm correct, things should succeed again with only logging memory leaks.

OK, the `SIGABRT`s are gone and the leaks are back, but this is not surprising at all, because this is exactly the state in which it was before I started calling the destructor manually. And the memory leaks are back, so the destructor obviously isn't called at all when the destructor fails with an exception, which is also the expected behavior as described in the C++ specs. That's why I added the manual destructor call.

So we're back at the beginning. What now? How to resolve the memory leaks?

If destructor is considered so special in C++ that invoking it manually also unconditionally tries to delete some random members of the object, should I extract the pointer deletions to a custom method which already can be called from both CTOR (if the `_read()` inside it fails) and DTOR?

**Comment by ams-tschoening** [comment]
> So we're back at the beginning. What now? How to resolve the memory leaks?

Not really at the beginning, what you did is still necessary. :-) An additional non-DTOR-cleanup is one way, but I would start with asking a question at Stackoverflow about the concrete behaviour we see here. The interesting point is if CLANG in the version you use is expected to cleanup the `std::string` two times in this scenario, if that is even C++ expected behaviour in general and stuff like that. You mentioned to didn't see those problems with other compilers.

So let's first try to understand what the actual problem is with the current implementation, CLANG and automatically cleaning things up in DTOR and CTOR and stuff. I would ask that question out of interest on SO anyway, but not today anymore. So either you wait until I do tomorrow or ask yourself.

**Comment by generalmimon** [comment]
@ams-tschoening:

> I would ask that question out of interest on SO anyway, but not today anymore. So either you wait until I do tomorrow or ask yourself.

That would be awesome, after all you know a lot more about C++ than I do, so you can prepare some minimal sample for SO purposes and formulate the problem more easily than me 😃

> So either you wait until I do tomorrow or ask yourself.

Tomorrow it will be 2.5 years since 0.8 release, so I think that a few days don't matter 🙂

**Comment by ams-tschoening** [comment]
Let's see how PROs explain this behaviour and the differences between the compilers:

https://stackoverflow.com/questions/63215793/why-does-clang-3-5-on-linux-cleans-a-stdstring-up-twice-when-calling-dtor-wh

@generalmimon It might be a good idea to at least test how things behave with a custom cleanup method instead of calling the DTOR manually. I would just suggest not calling it `destruct`, but really something more concrete like `cleanUp` or alike. There a re lot of `init`s implementing heavy work out there as well.

**Comment by generalmimon** [comment]
> Let's see how PROs explain this behaviour and the differences between the compilers:
> 
> https://stackoverflow.com/questions/63215793/why-does-clang-3-5-on-linux-cleans-a-stdstring-up-twice-when-calling-dtor-wh

Thanks a lot for asking that question on StackOverflow! Hopefully someone knows what's going on.

> @generalmimon It might be a good idea to at least test how things behave with a custom cleanup method instead of calling the DTOR manually. I would just suggest not calling it `destruct`, but really something more concrete like `cleanUp` or alike. There a re lot of `init`s implementing heavy work out there as well.

I've just pushed the change, and it finally fixed all the memory leaks! 🎉 There really seems to be something magical about calling the destructor manually.

@ams-tschoening Thanks a lot for your help! I think I can close this now, because all memory leaks relating to this issue has been fixed, according to the [CI dashboard](http://ci.kaitai.io/).

If we get some answers on SO which would led to some improvements of the current approach, I can apply them. But as I said, the issue is already resolved.

**Comment by ams-tschoening** [comment]
> Let's see how PROs explain this behaviour and the differences between the compilers:

In the end, things are pretty easy: DTORs are expected to simply [always cleanup an object](https://stackoverflow.com/a/63231169/2055163) and that includes members:

> **[class.dtor]/14** After executing the body of the destructor and destroying any objects with automatic storage duration allocated within the body, **a destructor for class X calls the destructors for X’s direct non-variant non-static data members**, the destructors for X’s non-virtual direct base classes and, if X is the most derived class (11.10.2), its destructor calls the destructors for X’s virtual base classes...

CTORs and DTORs are expected to [cleanup if they terminate unexpectedly](https://stackoverflow.com/a/35858302/2055163):

> C++11 15.2 Constructors and destructors (2)

> An object of any storage duration whose **initialization or destruction is terminated by an exception** will have destructors executed for all of its fully constructed subobjects (excluding the variant members of a union-like class), that is, for subobjects for which the principal constructor (12.6.2) has completed execution and the destructor has not yet begun execution.

`subobject` is the phrase used for [data members](https://stackoverflow.com/a/18452260/2055163) in this case:

> **[intro.object]/2** "Objects can contain other objects, called subobjects. A subobject can be a member subobject, a base class subobject, or an array element. An object that is not a subobject of any other object is called a complete object."

So, calling the DTOR manually in case of an exception in the CTOR simply is wrong in most of the cases, because if this works or not heavily depends on data members and stuff. Why other compilers didn't aborted with similar errors like CLANG 3.5 on Linux is pure coincidence and can't be relied upon. Having an additional method to clean dynamically maintained memory up is the correct thing to do in this case, unless one has a concept like smart pointers available. Those would be handled by the cleanup happening automatically when throwing in the CTOR, while the DTOR wouldn't be called manually most likely anyway. If it was, that would again be invalid like it was for `std::string`.

While all of this makes sense, I thought the automatic cleanup conceptually takes place elsewhere, outside at least the DTOR for data members and wasn't sure about throwing CTORs as well.

**Comment by generalmimon** [comment]
@ams-tschoening Thanks a lot for the explanation!

If the destructor is really meant to _destroy_ the object and _"any further operation on it is UB"_ anyway (https://stackoverflow.com/a/63219170), does it mean that we don't need anything like safeguards or setting the deleted pointers to `NULL`? AFAIK the main reason why we did these things was to allow multiple calls of the DTOR, but if that might break even if the DTOR body is empty (as calling the DTOR actually _destructs_ the object, and destructing it for the second time is UB anyway), I don't see a reason why we should keep the safeguards. They only add additional complexity to the compiler and the generated code and make the generated C++ code less readable.

**Comment by webbnh** [comment]
As I understand it, the DTOR will be called when the exception propagates out of the CTOR, and the DTOR will clean up any subobjects.  However, it won't, by default, clean up anything which was dynamically allocated in the CTOR.  We have added code to the DTOR to handle that, but doing so led to the temptation to call the DTOR explicitly as part of handling the exception, and that proved disastrous.  Also, it led to adding the safeguards to try to avoid double-deallocations.

There seem (to me) to be two options for rectifying the situation.  One is to switch to a RAII model, such that there are no dynamic allocations or dynamic allocations are wrapped in an object (e.g., use `std::shared_ptr`, even though we have no intention of actually sharing the objects).  The other is to move the cleanup to a subroutine which either the CTOR or the DTOR can call, but which functions correctly when called twice.  (For this we either need safeguards of some sort, or we need to ensure that the pointers are set to null when they don't point to anything and then rely on `delete` handling that appropriately.)

**Comment by ams-tschoening** [comment]
> [...]does it mean that we don't need anything like safeguards or setting the deleted pointers to NULL?

No, those safeguards ARE needed! Just look at your `_clean_up`-method, it is always executed in case of an exception in the CTOR. But how do you know that e.g. the called `_read` succeeded or not? You can't without looking at the pointer `_read` writes data to. `_read` might fail already BEFORE writing that data, so BEFORE actually allocating a new `std::vector`. In that case the pointer MUST NOT be freed and that is what the safeguards assures.

OTOH, `_read` might fail later, AFTER a new `std::vector` has been allocated and in that case the pointer NEEDS to be freed. How does one distinguish? Using the safeguard, because the pointer is either `NULL` and can be ignored or not and needs to be freed. Additionally setting to `NULL` after freeing is necessary as well, because `cleanup` in theory might be called multiple times and this is the only way to make sure that `free` is really only applied once.

You already call `_clean_up` twice, once in the `catch` of the CTOR and once in the DTOR. While in theory the DTOR is NOT called this way anymore in case the CTOR throws, there's no reason to simply make extra sure that multiple calls are safe. Safeguards, setting to `NULL` and calling `_clean_up` in the CTOR in case of exceptions and in the DTOR always are the correct things to do in this case. And it seems to work right now... :-)

> As I understand it, the DTOR will be called when the exception propagates out of the CTOR[...]

No, that's exactly what does NOT happen. :-) The DTOR always cleans up on its own, because that what's a DTOR expected to do, and the CTOR cleans up on its own in this case, because it's left by an exception and in that case it's expected to clean things up as well. It doesn't call the DTOR in this case, though, but really cleans up on it's own. And we broke that in the past by manually calling the DTOR.

**Comment by generalmimon** [comment]
@ams-tschoening:

> But how do you know that e.g. the called `_read` succeeded or not? You can't without looking at the pointer `_read` writes data to.

> How does one distinguish? Using the safeguard, because the pointer is either `NULL` and can be ignored or not and needs to be freed.

The nice thing about it is that we don't have to distinguish manually, because `delete` operator already does it for us. As already [pointed out](https://github.com/kaitai-io/kaitai_struct/issues/214#issuecomment-667017261) by @webbnh, deleting a `NULL` pointer is guaranteed to do nothing ([source](https://en.cppreference.com/w/cpp/memory/new/operator_delete#:~:text=In%20all%20cases,do%20nothing.)):

> In all cases, if `ptr` is a null pointer, the standard library deallocation functions do nothing.

Why should we have additional safeguard if the same _safeguard_ is contained in the `delete` operator itself?

In fact, we need the safeguard only for deleting items of `std::vector`, because we're manually iterating it in `_clean_up()` method, so we'd likely encounter a `NullPointerException` (using the Java name, I assume C++ has something similar) without the safeguard.

> Additionally setting to `NULL` after freeing is necessary as well, because `cleanup` **in theory might be called multiple times** and this is the only way to make sure that `free` is really only applied once.

 Can you explain how exactly can be `_clean_up()` called multiple times without causing _undefined behavior_ in the C++ compiler? After we make it `private` to make sure nobody can mess up the object from outside, it can be called only once either directly from the CTOR when the `try..catch` wrapping the `_read()` catches an exception, or from DTOR when the object gets destructed after it's properly initialized.

AFAIK, there's no way how to "call CTOR" multiple times on the same object without getting UB. If the CTOR fails with an exception, you get no object anyway, so deleting that _no object_ also doesn't make sense.

And I think that I don't have to repeat that the custom DTOR body is not automatically called when the CTOR fails.

And yeah, I think we can expect the same `InvalidRead` in the `clang` compiler (which is in fact a materialization of UB) from this snippet as we were getting when calling the DTOR directly from the CTOR before:

```cpp
default_endian_expr_exception_t::doc_t* d = new default_endian_expr_exception_t::doc_t(&io);
d->~doc_t();
d->~doc_t();
```

or

```cpp
delete d;
delete d;
```

As we already found out, if the object `d` contains some _maintained_ members like the `std::string m_indicator`, this will try to free the `m_indicator` twice (which is UB, if I understand it correctly). It doesn't matter if the custom DTOR is empty at all or if it has/hasn't safeguards, because it would anyway end on that `m_indicator`. 

> there's no reason to simply make extra sure that multiple calls are safe.

As long as C++ compilers will automatically clean up the subobjects on every DTOR call, we **can't** make them safe, even if we would like to. And I think that one can't exactly _call_ DTOR multiple times on the same object anyway, because as the answers on SO were saying, DTOR is meant to literally **destroy** the object. It's gone. Nobody knows what happens when you try to do something with the variable that used to hold it, because the object no longer exists. 

I can't figure out any valid case when the `_clean_up()` could be called multiple times. If you can, please share it. I just don't want our C++ codegen to handle cases which don't exist, as there is no benefit from doing that.

**Comment by ams-tschoening** [comment]
> In fact, we need the safeguard only for deleting items of std::vector[...]

So what exactly is your point when you admit that those safeguards are necessary? It makes absolutely no sense to remove those safeguards in favor of something focusing on cleaning up the elements of `std::vector` only. The current approach is simply ALWAYS safe, whereas everything else MIGHT be safe. It heavily depend on how you change things.

> Can you explain how exactly can be _clean_up() called multiple times without causing undefined behavior in the C++ compiler?[...]

The current implementation of `_clean_up` doesn't ever cause undefined behaviour, simply because the safeguards are there and no additional magic is taking place. This is different to what happens in the DTOR or when throwing in the CTOR.

Besides that, the fact that it's private NOW and only called twice NOW doesn't guarantee that it's not called additional times in future for any reason. It simply might become public for some reason or one simply might add additional internal calls for some reason. Who knows? The point is that the CURRENT implementation is safe to be called as often as one wants and if you change that, you simply risk introducing undefined behaviour again.

For what benefit? Which problem do you actually want to solve right now?

All tests pass, the implementation is safe, it doesn't leak any memory anymore, it's not even too bloated or whatever. If you really want to change things, implement a concept of smart pointers to be used by C++98 and migrate to that.

> [...]I just don't want our C++ codegen to handle cases which don't exist, as there is no benefit from doing that.

You need safeguards to handle `std::vector` and alike anyway already. Go ahead and feel free to suggest another implementation to deal with those cases. Chances are 50:50 that you are hunting undefined behaviour in some compiler again... :-)

**Comment by generalmimon** [comment]
@ams-tschoening Fair enough, I just hoped that you have some clear situation which would indispensably require safeguards, when you used such strong words like ARE NEEDED, MUST etc. But if you advise to keep them just as prevention for future robustness, I get it. I only didn't see _any_ reason before, and I personally needed some 🙂 

> > In fact, we need the safeguard only for deleting items of std::vector[...]
> 
> So what exactly is your point when you admit that those safeguards are necessary? It makes absolutely no sense to remove those safeguards in favor of something focusing on cleaning up the elements of `std::vector` only.

In fact, it occurred to me in the middle of writing the comment 😆

Let's focus on other cool stuff, after all, this problem is solved as you said, and it uses a pretty solid and foolproof implementation 👍

Thank you again for your help!

**Comment by generalmimon** [comment]
One thing that's left to do here:

[ ] Mark `_clean_up()` method as `private` (it's only an internal implementation helper, so it should be excluded from the public API)
